src/components/anchor-navigation/anchor-navigation.component.tsx

```diff
@@ -0,0 +1,234 @@
+import React, {
+  useState,
+  useRef,
+  useEffect,
+  useMemo,
+  useCallback,
+  createRef,
+} from "react";
+import { isFragment } from "react-is";
+import invariant from "invariant";
+import throttle from "lodash/throttle";
+
+import Event from "../../__internal__/utils/helpers/events";
+import {
+  StyledAnchorNavigation,
+  StyledNavigation,
+  StyledContent,
+} from "./anchor-navigation.style";
+import AnchorNavigationItem, {
+  AnchorNavigationItemProps,
+} from "./anchor-navigation-item/anchor-navigation-item.component";
+
+export interface AnchorNavigationProps {
+  /** Child elements */
+  children?: React.ReactNode;
+  /** The AnchorNavigationItems components to be rendered in the sticky navigation.
+  It is important to maintain proper structure.
+  List of AnchorNavigationItems has to be wrapped in React.Fragment */
+  stickyNavigation?: React.ReactNode;
+}
+
+const SECTION_VISIBILITY_OFFSET = 200;
+const SCROLL_THROTTLE = 100;
+
+const AnchorNavigation = ({
+  children,
+  stickyNavigation,
+}: AnchorNavigationProps): JSX.Element => {
+  invariant(
+    isFragment(stickyNavigation),
+    "`stickyNavigation` prop in `AnchorNavigation` should be a React Fragment."
+  );
+
+  const hasCorrectItemStructure = useMemo(() => {
+    const incorrectChild = React.Children.toArray(
+      stickyNavigation.props.children
+    ).find((child: React.ReactNode) => {
+      return (
+        !React.isValidElement(child) ||
+        (child.type as React.FunctionComponent).displayName !==
+          "AnchorNavigationItem"
+      );
+    });
+
+    return !incorrectChild;
+  }, [stickyNavigation]);
+
+  invariant(
+    hasCorrectItemStructure,
+    `\`stickyNavigation\` prop in \`AnchorNavigation\` should be a React Fragment that only contains children of type \`${AnchorNavigationItem.displayName}\``
+  );
+
+  const [selectedIndex, setSelectedIndex] = useState(0);
+
+  const sectionRefs = useRef<React.RefObject<HTMLElement>[]>(
+    React.Children.map(
+      stickyNavigation.props.children,
+      (
+        child: React.ReactElement<
+          AnchorNavigationItemProps,
+          typeof AnchorNavigationItem
+        >
+      ) => child.props.target
+    )
+  );
+
+  const anchorRefs = useRef(
+    Array.from(
+      {
+        length: React.Children.count(stickyNavigation.props.children),
+      },
+      () => createRef<HTMLAnchorElement>()
+    )
+  );
+
+  const contentRef = useRef<HTMLDivElement>(null);
+
+  const navigationRef = useRef<HTMLUListElement>(null);
+
+  const isUserScroll = useRef(true);
+
+  const isUserScrollTimer = useRef<NodeJS.Timeout>();
+
+  const setSelectedAnchorBasedOnScroll = useCallback(() => {
+    // istanbul ignore if
+    // function is called only after component is rendered, so ref cannot hold a null value
+    if (navigationRef.current === null) return;
```

Comment:
you could probably wrap the whole function in `if (navigationRef.current !== null) { ...}` and ignore else here, again not critical